letsencrypt/client.py

```diff
@@ -354,57 +354,82 @@ def deploy_certificate(self, domains, privkey_path,
         with error_handler.ErrorHandler(self._rollback_and_restart, msg):
             # sites may have been enabled / final cleanup
             self.installer.restart()
-
-    def enhance_config(self, domains, redirect=None):
+    def enhance_config(self, domains, config):
         """Enhance the configuration.
 
-        .. todo:: This needs to handle the specific enhancements offered by the
-            installer. We will also have to find a method to pass in the chosen
-            values efficiently.
-
         :param list domains: list of domains to configure
 
-        :param redirect: If traffic should be forwarded from HTTP to HTTPS.
-        :type redirect: bool or None
+        :ivar config: Namespace typically produced by
+            :meth:`argparse.ArgumentParser.parse_args`.
+            it must have the redirect, hsts and uir attributes.
+        :type namespace: :class:`argparse.Namespace`
 
         :raises .errors.Error: if no installer is specified in the
             client.
 
         """
+
         if self.installer is None:
             logger.warning("No installer is specified, there isn't any "
                            "configuration to enhance.")
             raise errors.Error("No installer available")
 
+        if config is None:
+            logger.warning("No config is specified.")
+            raise errors.Error("No config available")
+
+        redirect = config.redirect
+        hsts = config.hsts
+        uir = config.uir # Upgrade Insecure Requests
+
         if redirect is None:
             redirect = enhancements.ask("redirect")
 
-        # When support for more enhancements are added, the call to the
-        # plugin's `enhance` function should be wrapped by an ErrorHandler
         if redirect:
-            self.redirect_to_ssl(domains)
+            self.apply_enhancement(domains, "redirect")
```

Comment:
Each of these calls to `apply_enhancement` should be wrapped in a `try: catch PluginError` block that suppresses the exception that is raised if the enhancement is already in place.

Otherwise, if a domain already redirects from HTTP to HTTPS, and the user runs `letsencrypt install --redirect --uir --hsts`, this call to `apply_enhancement` will raise `PluginError: Let's Encrypt has already enabled redirection` and not continue to apply the other enhancements.